File: android/nestopia/src/main/kotlin/net/sourceforge/nestopia/NestopiaSaveLocations.kt
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

package net.sourceforge.nestopia

import java.io.File
import java.security.MessageDigest

/** Absolute locations of the persistent files belonging to one game. */
internal data class NestopiaSaveLocations(
    /** Directory holding every persistent file of the session. */
    val directory: File,
    /** Cartridge battery/EEPROM contents, written by the emulated machine. */
    val battery: File,
    /** Automatic save state, written and restored by the wrapper. */
    val autosave: File,
)

/**
 * Derives save file locations from a stable game identity.
 *
 * The naming rule is mirrored by `NestopiaSaveLocations.swift` on Apple platforms
 * so both platforms address the same game with the same file name.
 */
internal object NestopiaSaveLocator {
    const val BATTERY_EXTENSION = "sav"
    const val AUTOSAVE_EXTENSION = "auto.nst"

    /** Hex characters of the ROM digest kept in a file name. */
    const val DIGEST_LENGTH = 16

    /** Longest sanitized display name kept in a file name. */
    const val DISPLAY_NAME_LIMIT = 48
    const val FALLBACK_DISPLAY_NAME = "game"

    /** Replaces everything outside `A-Z a-z 0-9 . _ -` so the name is safe on every file system. */
    fun sanitized(name: String): String {
        val builder = StringBuilder()
        for (character in name) {
            val allowed =
                character in 'a'..'z' ||
                    character in 'A'..'Z' ||
                    character in '0'..'9' ||
                    character == '.' ||
                    character == '_' ||
                    character == '-'
            if (allowed) {
                builder.append(character)
            } else if (builder.lastOrNull() != '-') {
                builder.append('-')
            }
        }
        val trimmed = builder.toString().take(DISPLAY_NAME_LIMIT).trim('-', '.')
        return trimmed.ifEmpty { FALLBACK_DISPLAY_NAME }
    }

    /** Combines a readable name with a content digest so renamed or moved ROMs keep their saves. */
    fun identity(
        displayName: String,
        digest: String?,
    ): String {
        val name = sanitized(displayName)
        return if (digest.isNullOrEmpty()) name else "$name-${digest.take(DIGEST_LENGTH)}"
    }

    fun locations(
        directory: File,
        identity: String,
    ): NestopiaSaveLocations =
        NestopiaSaveLocations(
            directory = directory,
            battery = File(directory, "$identity.$BATTERY_EXTENSION"),
            autosave = File(directory, "$identity.$AUTOSAVE_EXTENSION"),
        )

    /** Lowercase SHA-1 of the ROM contents, or null when the file cannot be read. */
    fun digest(rom: File): String? =
        runCatching {
            val digest = MessageDigest.getInstance("SHA-1")
            rom.inputStream().use { stream ->
                val buffer = ByteArray(1 shl 16)
                while (true) {
                    val read = stream.read(buffer)
                    if (read <= 0) break
                    digest.update(buffer, 0, read)
                }
            }
            digest.digest().joinToString("") { "%02x".format(it.toInt() and 0xFF) }
        }.getOrNull()

    /** Moves the first existing legacy battery file onto [destination] when that file is missing. */
    fun migrateLegacyBattery(
        candidates: List<File>,
        destination: File,
    ): Boolean {
        if (destination.exists()) return false
        val legacy =
            candidates.firstOrNull { it.absolutePath != destination.absolutePath && it.isFile }
                ?: return false
        destination.parentFile?.mkdirs()
        if (legacy.renameTo(destination)) return true
        return runCatching {
            legacy.copyTo(destination, overwrite = false)
            legacy.delete()
            true
        }.getOrDefault(false)
    }
}

File: swift/Sources/Nestopia/NestopiaSaveLocations.swift
// Copyright (C) 2026 Dominic Opitz
// SPDX-License-Identifier: GPL-2.0-or-later

import CryptoKit
import Foundation

/// Controls the automatic save state written and restored for one Nestopia session.
public struct NestopiaAutosaveConfiguration: Sendable, Equatable {
    /// Seconds between two periodic autosaves.
    public static let defaultInterval: TimeInterval = 30
    /// Shortest interval honored by the engine, regardless of the configured value.
    public static let minimumInterval: TimeInterval = 5

    /// Restores the previous autosave on start and keeps writing it while the game runs.
    public var isEnabled: Bool
    /// Seconds between two periodic autosaves. Values below ``minimumInterval`` are raised.
    public var interval: TimeInterval

    public init(
        isEnabled: Bool = true,
        interval: TimeInterval = NestopiaAutosaveConfiguration.defaultInterval
    ) {
        self.isEnabled = isEnabled
        self.interval = interval
    }

    var resolvedInterval: TimeInterval {
        max(Self.minimumInterval, interval)
    }
}

/// Absolute locations of the persistent files belonging to one game.
struct NestopiaSaveLocations: Equatable {
    /// Directory holding every persistent file of the session.
    let directory: URL
    /// Cartridge battery/EEPROM contents, written by the emulated machine.
    let battery: URL
    /// Automatic save state, written and restored by the wrapper.
    let autosave: URL
}

/// Derives save file locations from a stable game identity.
///
/// The naming rule is mirrored by `NestopiaSaveLocations.kt` on Android so both
/// platforms address the same game with the same file name.
enum NestopiaSaveLocator {
    static let batteryExtension = "sav"
    static let autosaveExtension = "auto.nst"
    /// Hex characters of the ROM digest kept in a file name.
    static let digestLength = 16
    /// Longest sanitized display name kept in a file name.
    static let displayNameLimit = 48
    static let fallbackDisplayName = "game"

    /// Replaces everything outside `A-Z a-z 0-9 . _ -` so the name is safe on every file system.
    static func sanitized(_ name: String) -> String {
        var result = ""
        for scalar in name.unicodeScalars {
            let isAllowed =
                (scalar >= "a" && scalar <= "z") || (scalar >= "A" && scalar <= "Z")
                || (scalar >= "0" && scalar <= "9") || scalar == "." || scalar == "_"
                || scalar == "-"
            if isAllowed {
                result.unicodeScalars.append(scalar)
            } else if result.last != "-" {
                result.append("-")
            }
        }
        result = String(result.prefix(displayNameLimit))
        while let last = result.last, last == "-" || last == "." { result.removeLast() }
        while let first = result.first, first == "-" || first == "." { result.removeFirst() }
        return result.isEmpty ? fallbackDisplayName : result
    }

    /// Combines a readable name with a content digest so renamed or moved ROMs keep their saves.
    static func identity(displayName: String, digest: String?) -> String {
        let name = sanitized(displayName)
        guard let digest, !digest.isEmpty else { return name }
        return "\(name)-\(digest.prefix(digestLength))"
    }

    static func locations(directory: URL, identity: String) -> NestopiaSaveLocations {
        NestopiaSaveLocations(
            directory: directory,
            battery: directory.appendingPathComponent("\(identity).\(batteryExtension)"),
            autosave: directory.appendingPathComponent("\(identity).\(autosaveExtension)")
        )
    }

    /// Lowercase SHA-1 of the ROM contents, or `nil` when the file cannot be read.
    static func digest(forROMAt url: URL) -> String? {
        guard let handle = try? FileHandle(forReadingFrom: url) else { return nil }
        defer { try? handle.close() }
        var hasher = Insecure.SHA1()
        while let chunk = try? handle.read(upToCount: 1 << 16), !chunk.isEmpty {
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }

    /// Default directory used when the configuration does not name one.
    static func defaultDirectory() -> URL {
        let base =
            FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
        return base.appendingPathComponent("Nestopia/Saves", isDirectory: true)
    }

    /// Battery files written before saves were keyed by content digest.
    static func legacyBatteryURLs(romURL: URL, directory: URL) -> [URL] {
        let rawName = romURL.deletingPathExtension().lastPathComponent
        guard !rawName.isEmpty else { return [] }
        let fileName = "\(rawName).\(batteryExtension)"
        var candidates = [directory.appendingPathComponent(fileName)]
        guard
            let applicationSupport = FileManager.default.urls(
                for: .applicationSupportDirectory, in: .userDomainMask
            ).first
        else { return candidates }
        for legacyDirectory in ["Nestopia/Saves", "NES/Saves"] {
            candidates.append(
                applicationSupport
                    .appendingPathComponent(legacyDirectory, isDirectory: true)
                    .appendingPathComponent(fileName)
            )
        }
        return candidates
    }

    /// Moves the first existing legacy battery file onto `destination` when that file is missing.
    @discardableResult
    static func migrateLegacyBattery(from candidates: [URL], to destination: URL) -> Bool {
        let manager = FileManager.default
        guard !manager.fileExists(atPath: destination.path) else { return false }
        guard
            let legacy = candidates.first(where: {
                $0.standardizedFileURL != destination.standardizedFileURL
                    && manager.fileExists(atPath: $0.path)
            })
        else { return false }
        do {
            try manager.moveItem(at: legacy, to: destination)
            return true
        } catch {
            guard (try? manager.copyItem(at: legacy, to: destination)) != nil else { return false }
            try? manager.removeItem(at: legacy)
            return true
        }
    }

    /// Resolves, creates, and migrates the save locations for one session.
    static func resolve(romURL: URL, directory: URL?) throws -> NestopiaSaveLocations {
        let resolvedDirectory = directory ?? defaultDirectory()
        try FileManager.default.createDirectory(
            at: resolvedDirectory, withIntermediateDirectories: true)
        let locations = locations(
            directory: resolvedDirectory,
            identity: identity(
                displayName: romURL.deletingPathExtension().lastPathComponent,
                digest: digest(forROMAt: romURL)
            )
        )
        migrateLegacyBattery(
            from: legacyBatteryURLs(romURL: romURL, directory: resolvedDirectory),
            to: locations.battery
        )
        return locations
    }
}
